File: src/main/java/edu/uchicago/gerber/_08final/mvc/model/Asteroid.java
package edu.uchicago.gerber._08final.mvc.model;


import java.util.Comparator;
import java.util.LinkedList;
import java.util.Random;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Stream;
import java.awt.*;

import edu.uchicago.gerber._08final.mvc.controller.CommandCenter;
import edu.uchicago.gerber._08final.mvc.controller.Game;
import edu.uchicago.gerber._08final.mvc.controller.GameOp;
import edu.uchicago.gerber._08final.mvc.controller.SoundLoader;
import edu.uchicago.gerber._08final.mvc.model.prime.PolarPoint;


public class Asteroid extends Sprite {

	//radius of a large asteroid
	private final int LARGE_RADIUS = 110;

	//size determines if the Asteroid is Large (0), Medium (1), or Small (2)
	public Asteroid(int size){

		//a size of zero is a big asteroid
		//a size of 1 or 2 is med or small asteroid respectively. See getSize() method.
		if (size == 0) setRadius(LARGE_RADIUS);
		else setRadius(LARGE_RADIUS/(size * 2));

		//Asteroid is FOE
		setTeam(Team.FOE);
		setColor(Color.WHITE);

		//the spin will be either plus or minus 0-9
		setSpin(somePosNegValue(10));
		//random delta-x
		setDeltaX(somePosNegValue(10));
		//random delta-y
		setDeltaY(somePosNegValue(10));

		setCartesians(generateVertices());

	}



	//overloaded constructor, so we can spawn smaller asteroids from an exploding one
	public Asteroid(Asteroid astExploded){
		//calls the other constructor: Asteroid(int size)
		this(astExploded.getSize() + 1);
		setCenter((Point)astExploded.getCenter().clone());
		int newSmallerSize = astExploded.getSize() + 1;
		//random delta-x : inertia + the smaller the asteroid, the faster its possible speed
		setDeltaX(astExploded.getDeltaX() / 1.5 + somePosNegValue( 5 + newSmallerSize * 2));
		//random delta-y : inertia + the smaller the asteroid, the faster its possible speed
		setDeltaY(astExploded.getDeltaY() / 1.5 + somePosNegValue( 5 + newSmallerSize * 2));

	}

	//converts the radius to integer representing the size of the Asteroid:
	//0 = large, 1 = medium, 2 = small
	public int getSize() {
		int ratio = LARGE_RADIUS / getRadius(); // 1, 2, or 4
		//ratio = 1 → binary 0000 0001 → trailing zeros: 0
		//ratio = 2 → binary 0000 0010 → trailing zeros: 1
		//ratio = 4 → binary 0000 0100 → trailing zeros: 2
		return Integer.numberOfTrailingZeros(ratio);
	}



	private Point[] generateVertices() {


		final int VERTICES = 25 + Game.R.nextInt(7);    // 25–31 points
		final double MIN_RADIUS = 0.8;
		final double MAX_RADIUS = 1.0;

		Supplier<PolarPoint> polarPointSupplier = () -> {
			double r = MIN_RADIUS + (MAX_RADIUS - MIN_RADIUS) * Game.R.nextDouble();
			double theta = Game.R.nextDouble() * (2 * Math.PI);
			return new PolarPoint(r, theta);
		};

		Comparator<PolarPoint> byTheta = Comparator.comparingDouble(PolarPoint::getTheta);

		Function<PolarPoint, Point> toCartesian = pp -> {
			double radiusScale = 1000.0;
			int x = (int) (pp.getR() * radiusScale * Math.sin(pp.getTheta()));
			int y = (int) (pp.getR() * radiusScale * Math.cos(pp.getTheta()));
			return new Point(x, y);
		};

		return Stream.generate(polarPointSupplier)
				.limit(VERTICES)
				// Try removing the sorted call to see how asteroids render
				.sorted(byTheta)
				.map(toCartesian)
				.toArray(Point[]::new);
	}


	@Override
	public void draw(Graphics g) {
		renderVector(g);
	}

	@Override
	public void removeFromGame(LinkedList<Movable> list) {
		super.removeFromGame(list);
		spawnSmallerAsteroidsOrDebris();
		//give the user some points for destroying the asteroid
		CommandCenter.getInstance().setScore(CommandCenter.getInstance().getScore() + 10L * (getSize() + 1));

	}

	private void spawnSmallerAsteroidsOrDebris() {

		int size = getSize();
		//small (2) asteroids
		if (size > 1) {
			CommandCenter.getInstance().getOpsQueue().enqueue(new WhiteCloudDebris(this), GameOp.Action.ADD);
			SoundLoader.playSound("pillow.wav");
		}
		else {
			//for large (0) and medium (1) sized Asteroids only, spawn 2 or 3 smaller asteroids respectively
			//We can use the existing variable (size) to do this
			size += 2;
			while (size-- > 0) {
				CommandCenter.getInstance().getOpsQueue().enqueue(new Asteroid(this), GameOp.Action.ADD);
			}
			SoundLoader.playSound("kapow.wav");
		}

	}
}

File: src/main/java/edu/uchicago/gerber/_08final/mvc/model/Asteroid.kt
package edu.uchicago.gerber._08final.mvc.model

import edu.uchicago.gerber._08final.mvc.controller.Game
import edu.uchicago.gerber._08final.mvc.model.Movable.Team
import java.awt.Color
import java.awt.Graphics
import java.awt.Point
import java.util.function.BiFunction
import java.util.function.Supplier
import java.util.stream.Collectors
import java.util.stream.Stream

class Asteroid(size: Int) : Sprite() {
    //radius of a large asteroid
    companion object {
        private const val LARGE_RADIUS = 100
    }

    //size determines if the Asteroid is Large (0), Medium (1), or Small (2)
    //when you explode a Large asteroid, you should spawn 2 or 3 medium asteroids
    //same for medium asteroid, you should spawn small asteroids
    // get blasted into debris, but do not spawn anything
    init {

        //Asteroid is FOE
        team = Team.FOE

        color = Color.WHITE
        //the spin will be either plus or minus 0-9
        spin = somePosNegValue(10)
        //random delta-x
        deltaX = somePosNegValue(10).toDouble()
        //random delta-y
        deltaY = somePosNegValue(10).toDouble()

        //a size of zero is a big asteroid
        //a size of 1 or 2 is med or small asteroid respectively. See getSize() method.
        radius = if (size == 0) LARGE_RADIUS else LARGE_RADIUS / (size * 2)
        cartesians = genRandomPoints()
    }

    //overloaded so we can spawn smaller asteroids from an exploding one
    constructor(astExploded: Asteroid) : this(astExploded.size + 1) {
        center = astExploded.center
        val newSmallerSize = astExploded.size + 1
        //random delta-x : inertia + the smaller the asteroid, the faster its possible speed
        deltaX = astExploded.deltaX / 1.5 + somePosNegValue(5 + newSmallerSize * 2)
        //random delta-y : inertia + the smaller the asteroid, the faster its possible speed
        deltaY = astExploded.deltaY / 1.5 + somePosNegValue(5 + newSmallerSize * 2)
    }

    val size: Int
        get() = when (radius) {
            LARGE_RADIUS -> 0
            LARGE_RADIUS / 2 -> 1
            LARGE_RADIUS / 4 -> 2
            else -> 0
        }


    private fun genRandomPoints(): Array<Point> {
        //6.283 is the max radians
        val MAX_RADIANS_X1000 = 6283
        //when casting from double to int, we truncate and lose precision, so best to be generous with multiplier
        val PRECISION_MULTIPLIER = 1000;

        val polarPointSupplier = Supplier {
            val r = (800 + Game.R.nextInt(200)) / 1000.0 //number between 0.8 and 0.999
            val theta = Game.R.nextInt(MAX_RADIANS_X1000) / 1000.0 // number between 0 and 6.282
            PolarPoint(r, theta)
        }

        val transformer = BiFunction { spr: Sprite, pp: PolarPoint ->
            Point(
                (spr.center.x + pp.r * spr.radius * PRECISION_MULTIPLIER
                        * Math.sin(
                    Math.toRadians(spr.orientation.toDouble())
                            + pp.theta
                )).toInt(),
                (spr.center.y + pp.r * spr.radius * PRECISION_MULTIPLIER
                        * Math.cos(
                    Math.toRadians(spr.orientation.toDouble())
                            + pp.theta
                )).toInt()
            )
        }


        //random number of vertices
        val VERTICES = Game.R.nextInt(7) + 22
        return Stream.generate(polarPointSupplier)
            .limit(VERTICES.toLong())
            .sorted { pp1, pp2 -> pp1.theta.compareTo(pp2.theta) }
            .map { pp: PolarPoint -> transformer.apply(this, pp) }
            .collect(Collectors.toList())
            .toTypedArray()


    }
    override fun draw(g: Graphics) {
        renderVector(g)
    }


}
